src/binary-encoding.adoc

```diff
@@ -0,0 +1,1600 @@
+| `*`
+|Single-byte macro invocations
+
+|`0x4_`
+| `*`
+|Multibyte macro invocations
+
+.4+|`0x5_`
+| `0`-`8`
+|Integers
+
+|`9`
+<|Reserved
+
+|`A`-`D`
+<|Floats
+
+|`E`-`F`
+<|Booleans
+
+|`0x6_`
+|`*`
+|Decimals
+
+|`0x7_`
+|`*`
+|Timestamps
+
+|`0x8_`
+|`*`
+|Strings
+
+|`0x9_`
+|`*`
+|Symbols with inline text
+
+|`0xA_`
+|`*`
+|Lists
+
+|`0xB_`
+|`*`
+|S-expressions
+
+.3+|`0xC_`
+|`0`
+|Empty struct
+
+|`1`
+<|Reserved
+
+|`2`-`F`
+<|Structs with symbol address field names
+
+.2+|`0xD_`
+|`0`-`1`
+|Reserved
+
+|`2`-`F`
+<|Structs with `VarSym` field names
+
+.9+|`0xE_`
+|`0`
+|Ion version marker
+
+|`1`-`3`
+<|Symbols with symbol table text
+
+|`4`-`6`
+<|Annotations with symbol table text
+
+|`7`-`9`
+<|Annotations with `VarSym` text
+
+|`A`
+<|`null.null`
+
+|`B`
+<|Typed nulls
+
+|`C`-`D`
+<|NOP
+
+|`E`
+<|Reserved
+
+|`F`
+<|System macro invocation
+
+.16+|`0xF_`
+|`0`
+|Delimited container end
+
+|`1`
+<|Delimited list start
+
+|`2`
+<|Delimited S-expression start
+
+|`3`
+<|Delimited struct with `VarSym` field names start
+
+|`4`
+<|Variable length prefixed macro invocation
+
+|`5`
+<|Variable length integer
+
+|`6`
+<|Variable length decimal
+
+|`7`
+<|Variable length, long-form timestamp
+
+|`8`
+<|Variable length string
+
+|`9`
+<|Variable length symbol encoded as `VarSym`
+
+|`A`
+<|Variable length list
+
+|`B`
+<|Variable length S-expression
+
+|`C`
+<|Variable length struct with symbol address field names
+
+|`D`
+<|Variable length struct with `VarSym` field names
+
+|`E`
+<|Variable length blob
+
+|`F`
+<|Variable length clob
+
+|===
+
+
+
+[[single_byte_macro_invocations]]
+=== Single-byte macro invocations
+
+// TODO: link to macros chapter
+
+If the value of the opcode is less than `64` (`0x40`), it represents an invocation of the macro at the corresponding
+__address__—an offset within the local macro table.
+
+==== Example encoding of a single-byte invocation of the macro at address `7`
+[source]
+----
+0 0 0 0 0 1 1 1
+└──────┬──────┘
+  FixedUInt 7
+----
+
+==== Example encoding of an invocation of the macro at address `31`
+[source]
+----
+0 0 0 1 1 1 1 1
+└──────┬──────┘
+  FixedUInt 31
+----
+
+// TODO: Link to macro calling conventions
+
+Note that the opcode alone tells us which macro is being invoked, but it does not supply enough information for the
+reader to parse any arguments that may follow. The parsing of arguments is described in detail in the section _Macro
+calling conventions_.
+
+[[multi_byte_macro_invocations]]
+=== Multibyte macro invocations
+
+While invocations of macro addresses in the range `[0, 63]` can be encoded in a single byte using
+<<single_byte_macro_invocations, single byte macro invocations>>, many applications will benefit from defining more than
+64 macros.
+
+If the high nibble of the opcode is `0x4_`, then the low nibble represents the four least significant bits of the macro
+address. A <<varuint, `VarUInt`>> follows that contains the remaining, more significant bits.
```

Comment:
I'm not sure I follow; could you describe the arrangement you expected?

Reply:
> My interpretation is that the least four significant bits are in the op-code, followed by a VarUInt that encodes the remaining bits in LE order (each byte that follows modulo the encoding hold bits from lower to higher). So in principle the bit order is consistent, so I don't quite follow the "backwards" point as well.

This is what I was going for. 👍 I didn't move the continuation flag bits into the opcode's low nibble because it would require special machinery.